File: lib/websocket.js
const axios = require('axios');
const WebSocket = require('ws');
const { logger } = require('./log');
const { getSession } = require('./session');
const config = require('../config.json');

let ws;

async function enableWebsocket() {
  try {
    const enableWebsocketRes = await axios({
      method: 'post',
      url: '/config',
      baseURL: config.mirai.apiBaseURL,
      data: {
        sessionKey: getSession(),
        enableWebsocket: true,
      },
    });
    const { data: enableWebsocketData } = enableWebsocketRes;
    if (enableWebsocketData.code === 0) {
      logger.info('Enable websocket succeed.');
    } else {
      logger.warn('Enable websocket failed.');
    }
    logger.debug(enableWebsocketData);
  } catch (error) {
    logger.fatal(`Enable websocket failed. ${error.message}`);
    logger.debug(error);
    process.exit(-1);
  }
}

function onMessageHandle(data) {
  if (config.websocket.postURL) {
    axios({
      method: 'post',
      url: config.websocket.postURL,
      data: JSON.parse(data),
    })
      .then(function(res) {
        logger.info('Post data succeed.');
        logger.debug(res);
      })
      .catch(function(error) {
        logger.error(`Post data failed. ${error.message}`);
        logger.debug(error);
      });
  }
}

async function initWebsocket() {
  await enableWebsocket();
  ws = new WebSocket(`${config.websocket.baseURL}/all?sessionKey=${getSession()}`);
  ws.on('open', function() {
    logger.info('Websocket opened.');
  });
  ws.on('message', function(data) {
    logger.info('Receive message from websocket.');
    logger.debug(data);
    onMessageHandle(data);
  });
  ws.on('error', function(error) {
    logger.error('Websocket erred.');
    logger.debug(error);
  });
  ws.on('close', function(code, reason) {
    logger.fatal(`Websocket closed with code ${code} and reason '${reason}'.`);
    process.exit(-1);
  });
}

module.exports = {
  initWebsocket,
};

File: websocket.go
package main

import (
	"fmt"

	"github.com/go-resty/resty/v2"
	"github.com/gorilla/websocket"
	"github.com/rs/zerolog/log"
	"github.com/tidwall/gjson"
)

func enableWebsocket() {
	client := resty.New()
	client.SetCloseConnection(true)
	resp, err := client.R().
		SetHeader("Content-Type", "application/json").
		SetBody(map[string]interface{}{
			"sessionKey":      session,
			"enableWebsocket": true,
		}).
		Post(cfg.Get("mirai.apiBaseURL").String() + "/config")
	if err != nil {
		log.Error().Msgf("Enable websocket erred. %v", err)
		return
	}
	code := gjson.GetBytes(resp.Body(), "code")
	if code.Exists() && code.Value() == 0.0 {
		log.Info().Msg("Enable websocket succeed.")
		log.Debug().Msgf("%v", resp)
		return
	}
	log.Error().Msg("Enable websocket erred.")
	log.Debug().Msgf("%v", resp)
}

func startWebsocket(channel string) {
	for {
		if wsClosed {
			return
		}
		connectWebsocket(channel)
	}
}

func connectWebsocket(channel string) {
	if wsClosed {
		return
	}
	log.Info().Msgf(`Connecting websocket to channel "%v"...`, channel)
	wsConn, _, err := websocket.DefaultDialer.Dial(
		fmt.Sprintf(
			"%s%s?authKey=%s&sessionKey=%s",
			cfg.Get("mirai.wsBaseURL").String(),
			channel,
			cfg.Get("mirai.authKey").String(),
			session,
		),
		nil,
	)
	if err != nil {
		log.Error().Msgf(`Connect websocket to channel "%v" erred. %v`, channel, err)
		return
	}
	defer wsConn.Close()
	wsConns[channel] = wsConn
	log.Info().Msgf(`Websocket connected to channel "%v".`, channel)
	readWebsocket(channel)
}

func readWebsocket(channel string) {
	wsConn := wsConns[channel]
	for {
		messageType, message, err := wsConn.ReadMessage()
		if err != nil {
			if !wsClosed {
				log.Error().Msgf(`Listen websocket from channel "%v" erred. %v`, channel, err)
			}
			return
		}
		if messageType == websocket.TextMessage {
			log.Info().Msgf(`Receive websocket message from channel "%v".`, channel)
			log.Debug().Msgf("%s", message)
			messageHandler(message, channel)
		}
	}
}
